File: src/test/java/com/algo/linkedlist/PalindromeListTest.java
package com.algo.linkedlist;

import org.testng.annotations.Test;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;

public class PalindromeListTest {

    private final PalindromeList palindromeList = new PalindromeList(new ReverseList());

    @Test
    public void emptyListIsPalindrome() {
        assertThat(palindromeList.isPalindrome(null), is(true));
    }

    @Test
    public void singleNodeListIsPalindrome() {
        assertThat(palindromeList.isPalindrome(new ListNode(2)), is(true));
    }

    @Test
    public void List12IsNotPalindrome() {
        assertThat(palindromeList.isPalindrome(new ListNode(1, new ListNode(2))), is(false));
    }

    @Test
    public void List121IsPalindrome() {
        assertThat(palindromeList.isPalindrome(new ListNode(1, new ListNode(2, new ListNode(1)))), is(true));
    }
}

File: src/test/scala/com/algo/scala/linkedlist/PalindromeListTest.scala
package com.algo.scala.linkedlist

import com.aglo.scala.linkedlist.{EmptyLinkedList, LinkedList, ListNode, NonEmptyLinkedList}
import org.scalatest.FunSuite


class PalindromeListTest extends FunSuite {

  def emptyList: LinkedList = new EmptyLinkedList

  def singleNodeList(value: Int): LinkedList = new NonEmptyLinkedList(new ListNode(value), emptyList)

  def node(value: Int): ListNode = new ListNode(value)

  test("Empty List is Palindrome") {
    //given
    def empty = emptyList

    //when
    def palindrome = empty.isPalindrome

    //then
    assert(palindrome)
  }

  test("Single Node List is Palindrome") {
    //given
    def node = singleNodeList(1)

    //when
    def palindrome = node.isPalindrome

    //then
    assert(palindrome)
  }

  test("1 2 List is Not Palindrome") {
    //given
    def list = singleNodeList(2).addToHead(node(1))

    //when
    def palindrome = list.isPalindrome

    //then
    assert(!palindrome)
  }

  test("1 2 1 List is Palindrome") {
    //given
    def list = singleNodeList(1).addToHead(node(2)).addToHead(node(1))

    //when
    def palindrome = list.isPalindrome

    //then
    assert(palindrome)
  }

}
